src/audio_buffer.rs

```diff
@@ -0,0 +1,482 @@
+
+        for index in 0..max_frame {
+            if index < destination.len() {
+                destination[index] = channel[index + offset];
+            } else {
+                destination.push(channel[index + offset]);
+            }
+        }
+    }
+
+    // @note - not sure if we can handle default arguments in another way
+    pub fn copy_to_channel(&mut self, source: &[f32], channel_number: usize) {
+        self.copy_to_channel_with_offset(source, channel_number, 0);
+    }
+
+    pub fn copy_to_channel_with_offset(
+        &mut self,
+        source: &[f32],
+        channel_number: usize,
+        offset: usize,
+    ) {
+        // [spec] Let buffer be the AudioBuffer with 𝑁𝑏 frames, let 𝑁𝑓 be the number
+        // of elements in the source array, and 𝑘 be the value of bufferOffset. Then
+        // the number of frames copied from source to the buffer is max(0,min(𝑁𝑏−𝑘,𝑁𝑓)).
+        // If this is less than 𝑁𝑓, then the remaining elements of buffer are not modified.
+        //
+        // we use the `len` instead of `capacity` because source is a read buffer and we
+        // want to be sure we don't access some undefined index (cf. `copy_from_channel`)
+        let src_len = source.len();
+        let max_frame = (self.length - offset).min(src_len).max(0);
+        let channel = &self.internal_data.channels[channel_number];
+        // we need to copy the underlying channel here because it could have been
+        // acquired by some node and be in use.
+        // @see - https://webaudio.github.io/web-audio-api/#acquire-the-content
+        let mut copy = channel.to_vec();
+        copy[offset..(max_frame + offset)].copy_from_slice(&source[..max_frame]);
+
+        // replace channel with modified copy and new Arc so that next time some
+        // node acquire the content it will grab the updated values and the old
+        // data will be freed when last ref from audio node is dropped.
+        // The nodes who already acquired the Arc to the previous resource
+        // are therefore not impacted.
+        //
+        // @note - maybe this does not handle properly some edge case where:
+        //  `buffer is set to the source -> buffer is changed -> source is started`
+        //  but for most use-cases this seems to ok
+        let channels = Rc::make_mut(&mut self.internal_data.channels);
+        channels[channel_number] = Arc::new(copy);
+    }
+
+    // [spec] According to the rules described in acquire the content either allow writing
+    // into or getting a copy of the bytes stored in [[internal data]] in a new Float32Array
+    //
+    // @note - that's really not clear and kind of "do whatever you want unless it
+    // breaks something...", so just return a copy to make sure nothing can be messed up
+    pub fn get_channel_data(&self, channel_number: usize) -> Vec<f32> {
+        let channel = &self.internal_data.channels[channel_number];
+        channel.to_vec()
```

Comment:
Should we really copy the data? We can probably get away with your `get_channel_clone`, it will be way more efficient

Reply:
Or just return the full slice?

Reply:
It just to make sure nobody can mess up a buffer that is currently played. I would say that we could just keep things like that (possibly putting a warning in the doc), and discuss it more in depth inside the new issue about "acquire the content".